msal/src/test/java/com/microsoft/identity/client/e2e/tests/mocked/GetDeviceCodeTest.java

```diff
@@ -0,0 +1,155 @@
+        Assert.assertNotNull(authorizationResponse.getInterval());
+        Assert.assertNotNull(authorizationResponse.getExpiresIn());
+        Assert.assertNotNull(authorizationResponse.getVerificationUri());
+
+        Assert.assertNull(authorizationResult.getAuthorizationErrorResponse());
+    }
+
+    @Test
+    public void testGetDeviceCodeFailureNoClientId() {
+        final MicrosoftStsAuthorizationRequest authorizationRequest = builder.setClientId(null).build();
+        final AuthorizationResult authorizationResult = runGetDeviceCodeThread(authorizationRequest, urlBody);
+        final MicrosoftStsAuthorizationErrorResponse authorizationErrorResponse = (MicrosoftStsAuthorizationErrorResponse) authorizationResult.getAuthorizationErrorResponse();
+
+        Assert.assertFalse(authorizationResult.getSuccess());
+        Assert.assertNull(authorizationResult.getAuthorizationResponse());
+
+        Assert.assertNotNull(authorizationErrorResponse);
+        Assert.assertEquals("invalid_request", authorizationErrorResponse.getError());
+    }
+
+    @Test
+    public void testGetDeviceCodeFailureNoScope() {
+        final MicrosoftStsAuthorizationRequest authorizationRequest = builder.setScope(null).build();
+        final AuthorizationResult authorizationResult = runGetDeviceCodeThread(authorizationRequest, urlBody);
+        final MicrosoftStsAuthorizationErrorResponse authorizationErrorResponse = (MicrosoftStsAuthorizationErrorResponse) authorizationResult.getAuthorizationErrorResponse();
+
+        Assert.assertFalse(authorizationResult.getSuccess());
+        Assert.assertNull(authorizationResult.getAuthorizationResponse());
+
+        Assert.assertNotNull(authorizationErrorResponse);
+        Assert.assertEquals("invalid_request", authorizationErrorResponse.getError());
+    }
+
+    @Test
+    public void testGetDeviceCodeFailureBadScope() {
+        final MicrosoftStsAuthorizationRequest authorizationRequest = builder.setScope("/").build();
+        final AuthorizationResult authorizationResult = runGetDeviceCodeThread(authorizationRequest, urlBody);
+        final MicrosoftStsAuthorizationErrorResponse authorizationErrorResponse = (MicrosoftStsAuthorizationErrorResponse) authorizationResult.getAuthorizationErrorResponse();
+
+        Assert.assertFalse(authorizationResult.getSuccess());
+        Assert.assertNull(authorizationResult.getAuthorizationResponse());
+
+        Assert.assertNotNull(authorizationErrorResponse);
+        Assert.assertEquals("invalid_scope", authorizationErrorResponse.getError());
+    }
+
+    /**
+     * Helper function to run getDeviceCode(). Catches exception
+     * @param authorizationRequest request to send to getDeviceCode()
+     * @param urlBody url to send to getDeviceCode()
+     * @return authorizationResult from getDeviceCode()
+     */
+    private AuthorizationResult runGetDeviceCodeThread(final MicrosoftStsAuthorizationRequest authorizationRequest, final String urlBody) {
```

Comment:
what is this testing? why are we expecting IOException?

Reply:
This should be removed. I was originally running getDeviceCode in a separate thread each time but I realized that wasn't necessary when testing. This was a helper function that set up the thread and everything, it's not actually a test method. The try catch just made it fail if an exception was thrown.